# Exercise 4-7

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# import dataframe
ords_prods_merge = pd.read_pickle(r"C:\Users\jinu5\Desktop\careerfoundry\Achivement 4\Instacart Basket Analysis_2024-01-02\02 Data\Prepared Data\ords_prods_merge.pkl")

In [3]:
# create subsut dataframe of original dataframe, for escaping memory issues, etc. 
df=ords_prods_merge[:1000000]

In [4]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        0.0         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [5]:
# define funtion price_label(row)

def price_label(row):
    if row['prices'] <= 5: 
        return 'Low-range product'
    elif (row['prices']  > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range product'
    else: return 'Not enough data'

In [6]:
# add price label to each product in 'price_range' column

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\jinu5\AppData\Local\Temp\ipykernel_9732\2869886121.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [7]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [8]:
df['prices'].max()

14.8

In [9]:
# creating 'price_range_loc' byusing loc() function
df.loc[df['prices']>15, 'price_range_loc'] ='High-range product'

C:\Users\jinu5\AppData\Local\Temp\ipykernel_9732\482680839.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices']>15, 'price_range_loc'] ='High-range product'


In [10]:
df.head()


order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        0.0         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

  price_range_loc  
0             nan  
1             nan  
2             nan  
3             nan  
4             nan

In [11]:
df.loc [(df['prices']<=15) & (df['prices']>5), 'price_range_loc'] = 'Mid-range product'

In [12]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low_range_product'

In [13]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product    756450
Low_range_product    243550
Name: count, dtype: int64

In [14]:
# creating 'price_range_loc' byusing loc() function to ords_prods_merge dataframe

ords_prods_merge.loc[ords_prods_merge['prices']>15, 'price_range_loc'] ='High-range product'

In [15]:
ords_prods_merge.loc [(ords_prods_merge['prices' ]<=15) & (ords_prods_merge['prices']>5), 'price_range_loc'] = 'Mid-range product'

In [16]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low_range_product'

In [17]:
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low_range_product     10126321
High-range product      417678
Name: count, dtype: int64

In [18]:
for x in range (30,45):
    print ("my age is %a " %(x))

my age is 30 
my age is 31 
my age is 32 
my age is 33 
my age is 34 
my age is 35 
my age is 36 
my age is 37 
my age is 38 
my age is 39 
my age is 40 
my age is 41 
my age is 42 
my age is 43 
my age is 44 


In [19]:
# busiest day
ords_prods_merge['order_day_of_week'].value_counts(dropna=False)

order_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [20]:
result = []

for value in ords_prods_merge["order_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [21]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [22]:
ords_prods_merge['busiest_day']=result

In [23]:
ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# Task 4-7

### Step 2. 2 busiest days and 2 slowest days

In [ ]:
# 2 busiest days - 0 and 1 // slowest days - 3 and 4

busyslow =[]
for days in ords_prods_merge["order_day_of_week"]:
  if days == 0 or days==1 :
    busyslow.append("Busiest days")
  elif days == 4 or days==3:
    busyslow.append("Slowest days")
  else:
    busyslow.append("Regularly busy")

In [ ]:
# apply the label to 'busiest days'
ords_prods_merge['busiest_days']=busyslow

### Step 3. Check the values of this new column for accuracy & note observations

In [70]:
# value count
ords_prods_merge['busiest_days'].value_counts(dropna=False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [71]:
ords_prods_merge['busiest_days'].shape

(32404859,)

#### note
 - the counts of busiest dasys are some as order_day_of_week 0 + order_day_of_week 1. 
 - the counts of slowest dasys are some as order_day_of_week 3 + order_day_of_week 4. 
 - the total count is 32404859, which means every reconds have lables. 

### Step 4 identify the busiest hours of the day 
 - labeled “Most orders,” “Average orders,” and “Fewest orders.” 
 - Create a new column containing these labels called “busiest_period_of_day.”

In [72]:
# check frequency of order by hours
ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

#### criteria for categorizeing is not given. only 'too many users'
 - most orders: more than 2,500,000 (10, 11, 14, 15, 13, 12, 16)
 - fewest orders: less than 200,000 (1, 5, 2, 4, 3)

In [141]:
ords_prods_merge.loc [(ords_prods_merge['order_hour_of_day' ] >=10) & (ords_prods_merge['order_hour_of_day']<=16), 'busiest_period_of_day'] = 'Most orders'

In [142]:
ords_prods_merge.loc [(ords_prods_merge['order_hour_of_day' ] >=1) & (ords_prods_merge['order_hour_of_day']<=5), 'busiest_period_of_day'] = 'Fewest orders'

In [143]:
ords_prods_merge['busiest_period_of_day'].fillna('Average orders', inplace=True)


In [144]:
ords_prods_merge['busiest_period_of_day']

0                     nan
1                     nan
2             Most orders
3                     nan
4             Most orders
                ...      
32404854      Most orders
32404855      Most orders
32404856              nan
32404857    Fewest orders
32404858      Most orders
Name: busiest_period_of_day, Length: 32404859, dtype: object

In [145]:
ords_prods_merge['busiest_period_of_day'].replace('nan', 'Average orders', inplace=True)

In [146]:
ords_prods_merge['busiest_period_of_day']

0           Average orders
1           Average orders
2              Most orders
3           Average orders
4              Most orders
                 ...      
32404854       Most orders
32404855       Most orders
32404856    Average orders
32404857     Fewest orders
32404858       Most orders
Name: busiest_period_of_day, Length: 32404859, dtype: object

### Print the frequency of the busiest_period_of_day

In [147]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       18663868
Average orders    13363432
Fewest orders       377559
Name: count, dtype: int64

### Export dataframe

In [149]:
ords_prods_merge.to_pickle(r"C:\Users\jinu5\Desktop\careerfoundry\Achivement 4\Instacart Basket Analysis_2024-01-02\02 Data\Prepared Data\ords_prods_merge.pkl")